pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/ParentArgMinMaxAggregationFunction.java

```diff
@@ -0,0 +1,382 @@
+
+  public ParentArgMinMaxAggregationFunction(List<ExpressionContext> arguments, boolean isMax) {
+
+    super(arguments);
+    _isMax = isMax;
+    _functionIdContext = arguments.get(0);
+
+    _numMeasuringColumnContext = arguments.get(1);
+    _numMeasuringColumns = _numMeasuringColumnContext.getLiteral().getIntValue();
+
+    _measuringColumns = arguments.subList(2, 2 + _numMeasuringColumns);
+    _projectionColumns = arguments.subList(2 + _numMeasuringColumns, arguments.size());
+    _numProjectionColumns = _projectionColumns.size();
+  }
+
+  @Override
+  public AggregationFunctionType getType() {
+    return _isMax ? AggregationFunctionType.ARGMAX : AggregationFunctionType.ARGMIN;
+  }
+
+  @Override
+  public List<ExpressionContext> getInputExpressions() {
+    ArrayList<ExpressionContext> expressionContexts = new ArrayList<>();
+    expressionContexts.add(_functionIdContext);
+    expressionContexts.add(_numMeasuringColumnContext);
+    expressionContexts.addAll(_measuringColumns);
+    expressionContexts.addAll(_projectionColumns);
+    return expressionContexts;
+  }
+
+  @Override
+  public AggregationResultHolder createAggregationResultHolder() {
+    return new ObjectAggregationResultHolder();
+  }
+
+  @Override
+  public GroupByResultHolder createGroupByResultHolder(int initialCapacity, int maxCapacity) {
+    return new ObjectGroupByResultHolder(initialCapacity, maxCapacity);
+  }
+
+  @SuppressWarnings("LoopStatementThatDoesntLoop")
+  @Override
+  public void aggregate(int length, AggregationResultHolder aggregationResultHolder,
+      Map<ExpressionContext, BlockValSet> blockValSetMap) {
+
+    ArgMinMaxObject argMinMaxObject = aggregationResultHolder.getResult();
+
+    if (argMinMaxObject == null) {
+      initializeWithNewDataBlocks(blockValSetMap);
+      argMinMaxObject = new ArgMinMaxObject(_measuringColumnSchema.get(), _projectionColumnSchema.get());
+    }
+
+    List<Integer> rowIds = new ArrayList<>();
+    for (int i = 0; i < length; i++) {
+      int compareResult = argMinMaxObject.compareAndSetKey(_argMinMaxWrapperMeasuringColumnSets.get(), i, _isMax);
+      if (compareResult == 0) {
+        // same key, add the rowId to the list
+        rowIds.add(i);
+      } else if (compareResult > 0) {
+        // new key is set, clear the list and add the new rowId
+        rowIds.clear();
+        rowIds.add(i);
+      }
+    }
+
+    // for all the rows that are associated with the extremum key, add the projection columns
+    for (Integer rowId : rowIds) {
+      argMinMaxObject.addVal(_argMinMaxWrapperProjectionColumnSets.get(), rowId);
+    }
+
+    aggregationResultHolder.setValue(argMinMaxObject);
+  }
+
+  // this method is called to initialize the schemas if they are not initialized
+  // and to set the new block value sets for the wrapper classes
+  private void initializeWithNewDataBlocks(Map<ExpressionContext, BlockValSet> blockValSetMap) {
```

Comment:
this function is quite large and hard to read. can you modularize it a bit if possible?

Reply:
refactored